distributedcombigrid/src/sgpp/distributedcombigrid/manager/ProcessGroupWorker.cpp

```diff
@@ -570,6 +587,476 @@ void ProcessGroupWorker::combineUniform() {
+  // to be exchanged
+  // todo: use a flag to switch on/off optimized combination
+
+  reduceSparseGridCoefficients(lmax,lmin,combiParameters_.getNumberOfCombinations(),currentCombi_,
+      combiParameters_.getLMinReductionVector(), combiParameters_.getLMaxReductionVector());
+  /*for (size_t i = 0; i < lmax.size(); ++i)
+        if (lmin[i] > 1)
+          lmin[i] -= 01;
+  for (size_t i = 0; i < lmax.size(); ++i)
+      lmax[i] = std::max(lmin[i],lmax[i] - 2);
+  */
+#ifdef DEBUG_OUTPUT
+  MASTER_EXCLUSIVE_SECTION{
+    std::cout << "lmin: "<< lmin << std::endl;
+    std::cout << "lmax: "<< lmax << std::endl;
+  }
+#endif
+
+  //delete old dsgs
+  combinedUniDSGVector_.clear();
+  // erzeug dsgs
+  combinedUniDSGVector_.resize(numGrids);
+  for (auto& uniDSG : combinedUniDSGVector_) {
+    uniDSG = std::unique_ptr<DistributedSparseGridUniform<CombiDataType>>(
+        new DistributedSparseGridUniform<CombiDataType>(dim, lmax, lmin, boundary,
+                                                        theMPISystem()->getLocalComm()));
+  }
+  // todo: move to init function to avoid reregistering
+  // register dsgs in all dfgs
+  for (Task* t : tasks_) {
+    for(int g=0; g<numGrids; g++){
+
+      DistributedFullGrid<CombiDataType>& dfg = t->getDistributedFullGrid(g);
+
+      dfg.registerUniformSG(*(combinedUniDSGVector_[g]));
+    }
+  }
+  Stats::stopEvent("combine init");
+  Stats::startEvent("combine hierarchize");
+
+  real localMax(0.0);
+  //std::vector<CombiDataType> beforeCombi;
+  for (Task* t : tasks_) {
+    t->fullgridVectorBeforeCombi = new std::vector<CombiDataType>[numGrids];
+    for(int g=0; g<numGrids; g++){
+
+      DistributedFullGrid<CombiDataType>& dfg = t->getDistributedFullGrid(g);
+      std::vector<CombiDataType> datavector(dfg.getElementVector());
+      t->fullgridVectorBeforeCombi[g] = datavector;
+      
+      // hierarchize dfg
+      DistributedHierarchization::hierarchize<CombiDataType>(
+          dfg, combiParameters_.getHierarchizationDims() );
+      
+
+      // lokales reduce auf sg ->
+      dfg.addToUniformSG( *combinedUniDSGVector_[g], combiParameters_.getCoeff( t->getID() ) );
+#ifdef DEBUG_OUTPUT
+      std::cout << "Combination: added task " << t->getID() << " with coefficient " << combiParameters_.getCoeff( t->getID() ) <<"\n";
+#endif
+    }
+  }
+  Stats::stopEvent("combine hierarchize");
+
+  Stats::startEvent("combine global reduce init");
+
+  Task::bufAsync = new std::vector<CombiDataType>[numGrids];
+  Task::requestAsync = new MPI_Request[numGrids];
+
+  if(currentCombi_ == 0){
+    Task::subspaceSizes = new std::vector<int>[numGrids];
+    for(int g=0; g<numGrids; g++){
+      CombiCom::distributedGlobalReduceAsyncFirst( *combinedUniDSGVector_[g], Task::subspaceSizes[g]);
+    }
+  }
+
+  for(int g=0; g<numGrids; g++){
+    CombiCom::distributedGlobalReduceAsyncInit( *combinedUniDSGVector_[g], Task::subspaceSizes[g], Task::bufAsync[g], Task::requestAsync[g]);
```

Comment:
actually, I don't get why this is done once for each task -- a single time should suffice, right?

Reply:
it is only done a single time. Te numGrids does not refer to the number of tasks but to the number of grids each task owns. Usually numGrids is 1 but for example with multiple species in GENE numGrids=nspecies.